src/Catalog/Monitoring/Validation/Test/ValidationTestUtility.cs

```diff
@@ -0,0 +1,225 @@
+        /// </summary>
+        public class ValidationTestException : Exception
+        {
+            public ValidationTestException()
+                : base()
+            {
+            }
+
+            public ValidationTestException(string message)
+                : base(message)
+            {
+            }
+        }
+        
+        public class MetadataPropertiesInconsistencyException<T> : ValidationTestException
+        {
+            public MetadataPropertiesInconsistencyException(IEnumerable<PropertyInfo> inconsistentProperties, T v2Metadata, T v3Metadata)
+                : base($"The metadata between V2 and V3 is inconsistent! {string.Join(", ", inconsistentProperties?.Select(p => p.Name))} do not match!")
+            {
+                InconsistentProperties = inconsistentProperties;
+                V2Metadata = v2Metadata;
+                V3Metadata = v3Metadata;
+            }
+
+            public T V2Metadata { get; private set; }
+            public T V3Metadata { get; private set; }
+
+            public IEnumerable<PropertyInfo> InconsistentProperties { get; set; }
+        }
+
+        /// <summary>
+        /// We only verify these types. This prevents errors with unexpected fields.
+        /// </summary>
+        private static IDictionary<Type, IEqualityComparer<object>> _comparatorsForTypes = new Dictionary<Type, IEqualityComparer<object>>
+        {
+            { typeof(bool), new BoolComparer() },
+            { typeof(string), new StringComparer() },
+            { typeof(DateTimeOffset?), new DateTimeOffsetComparer() },
+            { typeof(NuGetVersion), new CastingComparer<NuGetVersion>() },
+            { typeof(Uri), new CastingComparer<Uri>() }
+        };
+
+        #region Comparers
+        /// <summary>
+        /// Returns true if both bools are equal.
+        /// Returns false otherwise.
+        /// </summary>
+        private class BoolComparer : IEqualityComparer<object>
+        {
+            public new bool Equals(object x, object y)
+            {
+                var boolX = (bool)x;
+                var boolY = (bool)y;
+
+                return boolX == boolY;
+            }
+
+            public int GetHashCode(object obj)
+            {
+                return obj.GetHashCode();
+            }
+        }
+
+        /// <summary>
+        /// Returns true
+        ///   if both <see cref="string"/>s are null or empty
+        ///   or if <see cref="string.Equals(string)"/> returns true.
+        ///   
+        /// Returns false otherwise.
+        /// </summary>
+        private class StringComparer : IEqualityComparer<object>
+        {
+            public new bool Equals(object x, object y)
+            {
+                var stringX = (string)x;
+                var stringY = (string)y;
+
+                if (string.IsNullOrEmpty(stringX))
```

Comment:
The main reason I added this was because for certain fields, the V2 implementation would return empty string and the V3 implementation would return null.